packages/react-utilities/src/compose/resolveShorthandProps.tsx

```diff
@@ -1,30 +1,28 @@
 import * as React from 'react';
+import { ObjectShorthandProps, ResolvedShorthandProps } from './types';
 
 /**
  * Ensures that the given slots are represented using object syntax. This ensures that
  * the object can be merged along with other objects.
  * @param props - The incoming props
  * @param shorthandPropNames - An array of prop names to apply simplification to
  */
-export const resolveShorthandProps = <TProps,>(props: TProps, shorthandPropNames: string[]) => {
+export const resolveShorthandProps = <TProps, TShorthandPropNames extends keyof TProps>(
+  props: TProps,
+  shorthandPropNames: readonly TShorthandPropNames[],
+): ResolvedShorthandProps<TProps, TShorthandPropNames> => {
   let newProps = props;
 
-  if (shorthandPropNames && shorthandPropNames.length) {
-    newProps = {
-      ...props,
-    };
-    for (const propName of shorthandPropNames) {
-      // TODO find clean way of guaranteeing only shorthand props are typechecked
-      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-      // @ts-ignore
-      const propValue = props[propName];
-
-      if (propValue !== undefined && (typeof propValue !== 'object' || React.isValidElement(propValue))) {
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        (newProps as any)[propName] = { children: propValue };
+  for (const propName of shorthandPropNames) {
+    const propValue = props[propName];
+    if (propValue !== undefined && (typeof propValue !== 'object' || React.isValidElement(propValue))) {
+      if (newProps === props) {
+        newProps = { ...props }; // Copy props before modifying
```

Comment:
Couldn't this potentially happen multiple times invalidating line 23 below? Shouldn't this be:

```suggestion
        newProps = { ...newProps }; // Copy props before modifying
```

Reply:
The goal was for it to only happen once, and I think line 19 guarantees that unless I'm missing something? Once it's copied, then `newProps === props` will be false, so it won't copy it again.

Reply:
The test on line 51 of `resolveShorthandProps.test.tsx` should verify that it works as expected to resolve multiple props without copying multiple times. If it copied the original `props` after resolving each shorthand prop, then it wouldn't work to resolve multiple shorthand props.